src/source.cpp

```diff
@@ -53,71 +54,103 @@ namespace librealsense
 
     void frame_source::init(std::shared_ptr<metadata_parser_map> metadata_parsers)
     {
-        std::lock_guard<std::mutex> lock(_callback_mutex);
+        _supported_extensions = { RS2_EXTENSION_VIDEO_FRAME,
+                                  RS2_EXTENSION_COMPOSITE_FRAME,
+                                  RS2_EXTENSION_POINTS,
+                                  RS2_EXTENSION_DEPTH_FRAME,
+                                  RS2_EXTENSION_DISPARITY_FRAME,
+                                  RS2_EXTENSION_MOTION_FRAME,
+                                  RS2_EXTENSION_POSE_FRAME };
 
-        std::vector<rs2_extension> supported { RS2_EXTENSION_VIDEO_FRAME,
-                                               RS2_EXTENSION_COMPOSITE_FRAME,
-                                               RS2_EXTENSION_POINTS,
-                                               RS2_EXTENSION_DEPTH_FRAME,
-                                               RS2_EXTENSION_DISPARITY_FRAME,
-                                               RS2_EXTENSION_MOTION_FRAME,
-                                               RS2_EXTENSION_POSE_FRAME };
+        _metadata_parsers = metadata_parsers;
+    }
 
-        for (auto type : supported)
-        {
-            _archive[type] = make_archive(type, &_max_publish_list_size, metadata_parsers);
-        }
+    void frame_source::create_archive( archive_id id )
+    {
+        auto it = std::find( _supported_extensions.begin(), _supported_extensions.end(), id.second );
```

Comment:
Consider protecting

Reply:
`_supported_extensions` is updated by `init` and `add_extension` and searched by `create_archive`.
Theoretically there could be a situation where the `add_extension` is called exactly when `create_archive` is running (e.g. called from `alloc_frame` or `begin_callback`).
I think that currently it does not happen in our code, but since I did not analyze all possible routes I preferred to protect.